querier/src/cache_system/backend/ttl.rs

```diff
@@ -0,0 +1,445 @@
+use std::{any::Any, fmt::Debug, hash::Hash, marker::PhantomData, sync::Arc, time::Duration};
+
+use time::{Time, TimeProvider};
+
+use super::{addressable_heap::AddressableHeap, CacheBackend};
+
+/// Interface to provide TTL (time to live) data for a key-value pair.
+pub trait TtlProvider: std::fmt::Debug + Send + Sync + 'static {
+    /// Cache key.
+    type K;
+
+    /// Cached value.
+    type V;
+
+    /// When should the given key-value pair expire?
+    ///
+    /// Return `None` for "never".
+    ///
+    /// The function is only called once for a newly cached key-value pair. This means:
+    /// - There is no need in remembering the time of a given pair (e.g. you can safely always return a constant).
+    /// - You cannot change the TTL after the data was cached.
+    ///
+    /// Expiration is set to take place AT OR AFTER the provided duration.
+    fn expires_in(&self, k: &Self::K, v: &Self::V) -> Option<Duration>;
+}
+
+/// [`TtlProvider`] that never expires.
+#[derive(Default)]
+pub struct NeverTtlProvider<K, V>
+where
+    K: 'static,
+    V: 'static,
+{
+    // phantom data that is Send and Sync, see https://stackoverflow.com/a/50201389
+    _k: PhantomData<fn() -> K>,
+    _v: PhantomData<fn() -> V>,
+}
+
+impl<K, V> std::fmt::Debug for NeverTtlProvider<K, V> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("NeverTtlProvider").finish_non_exhaustive()
+    }
+}
+
+impl<K, V> TtlProvider for NeverTtlProvider<K, V> {
+    type K = K;
+    type V = V;
+
+    fn expires_in(&self, _k: &Self::K, _v: &Self::V) -> Option<Duration> {
+        None
+    }
+}
+
+/// Cache backend that implements Time To Life.
+#[derive(Debug)]
+pub struct TtlBackend<K, V>
+where
+    K: Clone + Eq + Debug + Hash + Ord + Send + 'static,
+    V: Clone + Debug + Send + 'static,
+{
+    inner_backend: Box<dyn CacheBackend<K = K, V = V>>,
+    ttl_provider: Arc<dyn TtlProvider<K = K, V = V>>,
+    time_provider: Arc<dyn TimeProvider>,
+    expiration: AddressableHeap<K, (), Time>,
+}
+
+impl<K, V> TtlBackend<K, V>
+where
+    K: Clone + Eq + Debug + Hash + Ord + Send + 'static,
+    V: Clone + Debug + Send + 'static,
+{
+    /// Create new backend w/o any known keys.
+    ///
+    /// The inner backend SHOULD NOT contain any data at this point, otherwise we will not track any TTLs for these entries.
+    pub fn new(
+        inner_backend: Box<dyn CacheBackend<K = K, V = V>>,
+        ttl_provider: Arc<dyn TtlProvider<K = K, V = V>>,
+        time_provider: Arc<dyn TimeProvider>,
+    ) -> Self {
+        Self {
+            inner_backend,
+            ttl_provider,
+            time_provider,
+            expiration: Default::default(),
+        }
+    }
+
+    fn update(&mut self, now: Time) {
```

Comment:
Perhaps

```suggestion
    fn evict_expired(&mut self, now: Time) {
```

So usage below clearly shows what is happening:

```rust
fn get(&mut self, k: &Self::K) -> Option<Self::V> {
        self.evict_expired(self.time_provider.now());
        self.inner_backend.get(k)
    }
```

Could also inline acquiring the time from the time provider? Every call is `self.update(self.time_provider.now())`

Reply:
Renamed. Didn't inline it though, because we need the time from the time provider also during `set` and I wanted to avoid to call it twice.